lib/compress.js

```diff
@@ -320,7 +321,7 @@ merge(Compressor.prototype, {
                         value = node.fixed_value();
                         if (value && ref_once(d)) {
                             d.single_use = value instanceof AST_Lambda
```

Comment:
Instead of inserting `compressor` into `.is_constant_expression()` below, we could make this:
```js
compressor.options("reduce_funcs") && value instanceof AST_Lambda
```

Optional: pre-fetch `reduce_funcs` up here
https://github.com/mishoo/UglifyJS2/blob/94525d859f5db559366f96b646aca84438b0f982/lib/compress.js#L291

Reply:
> Instead of inserting compressor into .is_constant_expression() below, we could make this:
> `compressor.options("reduce_funcs") && value instanceof AST_Lambda`

But that makes 13 test cases produce larger code with 2 test failures. I probably misunderstood your instructions.

Would you mind just committing this PR and making the changes you like? It'd take less time.

Reply:
@alexlamsl Feel free to cancel this PR and make a better implementation of `reduce_funcs`. If you recycle this PR's tests you could save 10 minutes.

Reply:
I see what you mean - let me take a deeper look.

I hesitated because of the extensive changes to `.is_constant_expression()` :sweat:

Reply:
> If you recycle this PR's tests you could save 10 minutes.

I think your `README.md` changes would be best time-saver.

Reply:
FWIW, I do think `.is_constant_expression()` is the right place for the check.

Reply:
> I think your README.md changes would be best time-saver.

So my code is that useless? ;-)

Reply:
I meant I'm hopeless in writing documentation, if I ever remember to do so :sweat:

Reply:
Jokes aside, let's see if I understand what your requirements are - do you wish to disable things like `var o = { a: function() {} };` from being subsituted (assuming there is a single reference of `o` within the same scope)?

Reply:
> do you wish to disable things like var o = { a: function() {} }; from being subsituted (assuming there is a single reference of o within the same scope)?

Yeah, I saw that was caught up in this change as well. Since they are already function expressions, ideally it would be nice to permit that transform regardless of the `reduce_funcs` setting. `reduce_funcs` really only should gate `AST_Defun` to `AST_Function` substitution.

Reply:
Thanks for the clarification.

So you would like to see substitution of `var o = function() { ... }` disabled as well, regardless of whether `o` is within the same scope or not?

Reply:
I'm proposing to have `reduce_vars` and `collapse_vars` continue to transform `AST_Function` as it did prior to this PR - regardless of the `reduce_funcs` setting. So code like `var o = function() { ... }` would not produce another function expression instance - just merely move it, which is fine.